pkg/util/duration/duration.go

```diff
@@ -265,10 +265,61 @@ func Decode(sortNanos int64, months int64, days int64) (Duration, error) {
 
 // TODO(dan): Write DecodeBigInt.
 
-// Add returns the time t+d.
+// Add returns the time t+d, using the date-normalization behavior
+// implemented by PostgreSQL.
+//
+// go's implementation of time.Time.AddDate() doesn't match up with
+// the behavior we see in PostgreSQL when starting from days
+// numbered 29, 30, or 31 and advancing yearly or monthly.  Instead
+// of normalizing February 31st into the next month, we need to
+// "round down" to the end of the month.
```

Comment:
You have peeked into pg's implementation to determine what to do here. Please reference the particular pg version you looked at, and the filename inside the source tree you looked at, and the C function name(s) (or tests), in the comment.